Add rich support to torch.distributed.tensor.debug.visualize_sharding

wangkuiyi opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/151857

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

Please verify this PR by running the following command on a computer with at least 4 GPUs.

```shell
torchrun --nproc_per_node=4 /w/pytorch/torch/distributed/tensor/examples/visualize_sharding_example.py
```

wangkuiyi committed:
Update

wangkuiyi commented:
This PR is a WIP. I've tested the code as a standalone project. However, to make it part of PyTorch, I have to install PyTorch editable locally. When I was doing so on my MacBook following https://github.com/pytorch/pytorch#from-source, I got

```
09:59 $ python3 setup.py develop
Building wheel torch-2.8.0a0+gite31e2d2
-- Building version 2.8.0a0+gite31e2d2
Could not find any of CMakeLists.txt, Makefile, setup.py, LICENSE, LICENSE.md, LICENSE.txt in /Users/y/w/pytorch/third_party/psimd
Did you run 'git submodule update --init --recursive'?
```

I did run `git submodule update --init --recursive` and running it again didnt' help.

wangkuiyi committed:
Update

wangkuiyi commented:
@pytorchbot label "release notes: distributed (dtensor)"

wconstab commented on torch/distributed/tensor/debug/_visualize_sharding.py:
```diff
@@ -1,177 +1,251 @@
-
-
-def _create_table(blocks):
+def _shard_info(
```
I was glad to see `_compute_local_shape_and_global_offset` is deleted, but it seems like it is mostly copy-pasted into the new `_shard_info` fn.  If possible, i'd rather you find a way to call the real `torch.distributed._tensor._utils.compute_local_shape_and_global_offset` function, which I recently updated to handle uneven-sharding case.  This way the logic wouldn't diverge.  (I wasn't even aware of this copy ... )

wconstab commented on torch/distributed/tensor/debug/visualize_sharding_example.py:
```diff
@@ -0,0 +1,77 @@
+# OMP_NUM_THREADS=1 torchrun --nproc_per_node=4 visualize_sharding_example.py
```
why OMP_NUM_THREADS? i think for a visualization example this level of tuning should be unimportant?

wangkuiyi replied:
I included it solely to suppress the following warning.

```
W0423 20:24:18.427000 85647 site-packages/torch/distributed/run.py:792] *****************************************
W0423 20:24:18.427000 85647 site-packages/torch/distributed/run.py:792] Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed.
W0423 20:24:18.427000 85647 site-packages/torch/distributed/run.py:792] *****************************************
```

wconstab commented on torch/distributed/tensor/debug/visualize_sharding_example.py:
```diff
@@ -0,0 +1,77 @@
+import contextlib
+import importlib.util
+import os
+
```
for this example file, i'm fine landing it here for now, but i think longer term this might make sense to move the content into a 'tutorial' page on pytorch tutorials repo.

wangkuiyi replied:
Agreed.

wanchaol replied:
some duplicate files already in the examples folder, I think we should put it there first https://github.com/pytorch/pytorch/pull/152027#discussion_r2057401742

wconstab commented on torch/distributed/tensor/debug/visualize_sharding_example.py:
```diff
@@ -0,0 +1,77 @@
+    visualize_sharding(dt.DTensor.from_local(tensor, mesh, [dt.Shard(dim=1)]), use_rich=True)
+
+    # Case 2: 2D mesh
+    mesh = dist.init_device_mesh("cpu", (2, 2), mesh_dim_names=["dp", "tp"])
```
hmm. this example script is hanging for me.  I noticed that py-spy reports one rank is stuck on L24 barrier(), another rank is stuck on L44 init_device_mesh.

I also noticed that even though there were many calls to 'visualize_sharding' I only actually see one tabulate and one rich table.  
<img width="980" alt="image" src="https://github.com/user-attachments/assets/9bb1f756-c1e7-4dc8-8dad-b157307a0026" />

wangkuiyi replied:
I haven't been able to run it as part of PyTorch. I am still building PyTorch from source code on a Linux computer. I gave up building it on macOS.

Could you try:

```
pip install dtensor
torchrun --nproc_per_node=4 $(pip show dtensor | grep Location: | cut -f 2 -d ' ')/dtensor/[visualize_example.py](http://visualize_example.py/)
```

Here follows what I get. I don't see log message like `[Gloo] Rank 0 is connected to 3 peer ranks` as they are in your screenshot.

<img width="476" alt="Screenshot 2025-04-23 at 9 29 38 PM" src="https://github.com/user-attachments/assets/574d4eaf-c278-4137-bc0e-dd91bf7429aa" />

wconstab reviewed (commented):
Looks cool! i think there is something wrong though (see my note about the hang) - may need to take a deeper look.

wanchaol reviewed (commented):
Nice! have some comments inlined

wangkuiyi commented on torch/distributed/tensor/debug/visualize_sharding_example.py:
```diff
@@ -0,0 +1,77 @@
+
+
+@torch.distributed.elastic.multiprocessing.errors.record
+def main(local_device):
```
Ah! This makes total sense! Will do.

wangkuiyi commented on torch/distributed/tensor/debug/visualize_sharding_example.py:
```diff
@@ -0,0 +1,77 @@
+
+
+@contextlib.contextmanager
+def distributed_context():
```
Yes. Will remove this.

wangkuiyi commented on torch/distributed/tensor/debug/_visualize_sharding.py:
```diff
@@ -1,177 +1,251 @@
-                else:
-                    global_offset[shard_dim] += local_offset[shard_dim]
+def _has_rich_and_matplotlib() -> bool:
+    return importlib.util.find_spec("rich") and importlib.util.find_spec("matplotlib")
```
Users of rich do not have to install matplotlib.  We need matplitlib because it offers colormap, which is used to render neighboring cells in the table with different colors.

But you are right -- we should only try to import matplotlib if rich.console.Console tells us that the terminal does not support coloring. In that case, the current code would render tables using rich without coloring the cells.

I will make an edit soon.

wangkuiyi commented on torch/distributed/tensor/debug/_visualize_sharding.py:
```diff
@@ -1,177 +1,251 @@
-    .. note:: This requires the ``tabulate`` package. No sharding info will be printed for empty tensors
-    """
-    if dtensor.numel() == 0:  # we do not print for empty dtensors
+def visualize_sharding(dtensor, header="", use_rich: bool = False):
```
You are right - I forgot to remove this. Will remove it.

wangkuiyi replied:
Sorry - I just realized that we may need to keep this `use_rich`, so in the example, we could visualize each sharding with both the rich and the tabulate way and check their consistency. Does this make sense?

wanchaol replied:
I think maybe we should discuss about the expected behavior of this flag:
* if user installed rich/matplotlib and `use_rich=True`, they will use create_rich_table
* otherwise if user installed tabulate only, they will use tabulate

This means we could not by default let user to use the `rich` backend even if they installed it in their env.
I propose to change the argument `use_rich` to sth like `backend/render_backend:Literal["rich", "tabulate", "auto"]` and default to auto? This way:
* auto: auto select backend, prefer rich if installed
* the other two option: respect user choice and use that backend to render

lmk if that make sense :)

wanchaol replied:
discussed offline: we decided to keep the `use_rich=False`  by default to avoid any BC concerns.

wangkuiyi commented on torch/distributed/tensor/debug/_visualize_sharding.py:
```diff
@@ -1,177 +1,251 @@
-    """
-    if dtensor.numel() == 0:  # we do not print for empty dtensors
+def visualize_sharding(dtensor, header="", use_rich: bool = False):
+    """Visualizes sharding in the terminal for :class:`DTensor` that are 1D or 2D."""
```
Ah. I see. The use of`.. note::` instructs how PyTorch generate documentation, right? I will revert back to use it. Thank you for the advice!

wanchaol replied:
yeah mainly the docstring would be rendered in the documentation page.

wangkuiyi commented on torch/distributed/tensor/debug/_visualize_sharding.py:
```diff
@@ -1,177 +1,251 @@
+    ]
+
+    if _has_rich_and_matplotlib() and use_rich:
+        _create_rich_table(dtensor.shape, shards, device_kind=dtensor.device_mesh.device_type)
```
Would adding the header make the visualization different from the JAX style? If so, I guess we probably want to make sure that the layout, color, and other visual styles to be consistent with JAX's?

Indeed, I haven't figured out what print(headers) do? Let me try it.

wanchaol replied:
I honestly don't remember what's this headers for haha, and it's default to `""`, so it's fine to not print it.

wangkuiyi commented on torch/distributed/tensor/debug/_visualize_sharding.py:
```diff
@@ -1,177 +1,251 @@
+    return f"#{r:02X}{g:02X}{b:02X}"
+
+
+def _get_text_color(color: str) -> str:
```
The line chose a colormap provided by matplotlib named "tab20b".

```python
color_iter = make_color_iter(matplotlib.colormaps["tab20b"], num_rows, num_cols)
```

The color iterater chooses a color for each cell in the table.

So, I think there is no default color, but there is a default colormap.

wangkuiyi replied:
The bgcolor of a cell is chosen by `make_color_iter` from a colormap of matplotlib.

This function `_get_text_color` computes the font color from the bgcolor of a cell. If the bgcolor is dark, it returns a bright font color, otherwise, it returns a dark one.

wangkuiyi committed:
Adapt to run on 4 GPUs

wangkuiyi committed:
Update

wangkuiyi committed:
Update

wangkuiyi committed:
Update

wangkuiyi committed:
Use .. note:: in docstring

wangkuiyi committed:
Use torch/distributed/tensor/examples/visualize_sharding_example.py

wangkuiyi committed:
Remove unused print

wangkuiyi commented:
I have addressed all the comments, except [the one about `use_rich`](https://github.com/pytorch/pytorch/pull/152027#discussion_r2059501133).

Please take another look. Thanks!

wanchaol reviewed (approved):
lgtm, have one suggestion about `use_rich`, thanks for getting this to work!

wangkuiyi committed:
Update

wangkuiyi commented:
Fixed lint error.

wangkuiyi committed:
Remvoe | from type def

wconstab reviewed (approved):
nit: can you call `dist.destroy_process_group()` at the end of the script so it doesn't print this warning 

otherwise LGTM and runs fine now. thanks for doing this!

```
[rank1]:[W428 10:14:21.169010955 ProcessGroupNCCL.cpp:1556] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
```

wanchaol commented:
> nit: can you call `dist.destroy_process_group()` at the end of the script so it doesn't print this warning

@wconstab since this only lives in a example file and some small change, let's not trigger CI again. I can try to squeeze that change in some of my PRs.

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge -f "merge job time out"
